File: src/db/calc_GtEFbyET_new.R
setwd("~/github/CNEA-AQ/megan/src/db")

#Original Files
EF=read.csv("./orig/EFv210806.csv", sep=",")  

SpCrop=read.csv("./orig/SpeciationCrop210806.csv" , sep=",")  
SpHerb=read.csv("./orig/SpeciationHerb210806.csv" , sep=",")
SpShru=read.csv("./orig/SpeciationShrub210806.csv", sep=",")  
SpTree=read.csv("./orig/SpeciationTree210725.csv" , sep=","); SpTree=SpTree[,c(1:3)]  
#----------------------------------------------------------
SpCrop$Gtyp="Crop"
SpHerb$Gtyp="Herb"
SpShru$Gtyp="Shrub"
SpTree$Gtyp="Tree"
SpColumnNames=c("EcotypeID","VegID","SpFrac","Gtyp")

colnames(SpCrop)=SpColumnNames
colnames(SpHerb)=SpColumnNames
colnames(SpShru)=SpColumnNames
colnames(SpTree)=SpColumnNames

#Create new Speciation{Ntr,Btr}.csv files
df=merge(EF[c("VegID","GrowthForm")],SpTree, by="VegID",all.y = T)              #Add growthForm column

   #Needleleaf trees
   SpNtr=subset(df,GrowthForm=="Ntr")                                           #Get only Ntr species.
   NtEtFracs=aggregate(x = SpNtr$SpFrac, by=list(et=SpNtr$EcotypeID),FUN=sum)   #Compute partial-fraction: Ntr/tree
   for (i in c(1:nrow(NtEtFracs))){
     k=max(NtEtFracs$x[i],1e-23)
     et=NtEtFracs$et[i]
     SpNtr$SpFrac[which(SpNtr$EcotypeID==et)]=SpNtr$SpFrac[which(SpNtr$EcotypeID==et)]/k
   }
   #Broadleaf trees
   SpBtr=subset(df,GrowthForm=="Btr")                                           #Get only Btr species.
   BtEtFracs=aggregate(SpBtr$SpFrac,by=list(et=SpBtr$EcotypeID),FUN=sum)        #Compute partial fraction: Btr/Tree
   for (i in c(1:nrow(NtEtFracs))){
     k =max(1e-23,BtEtFracs$x[i])   
     et=BtEtFracs$et[i]   
     SpBtr$SpFrac[which(SpBtr$EcotypeID==et)]=SpBtr$SpFrac[which(SpBtr$EcotypeID==et)]/k   
   }
   ##I can write the new files:
   #write.csv(SpNtr[c("EcotypeID","VegID","SpFrac")],"./SpeciationNtree.csv")
   #write.csv(SpBtr[c("EcotypeID","VegID","SpFrac")],"./SpeciationBtree.csv")
   
SpNtr=SpNtr[SpColumnNames]; SpNtr$Gtyp="Ntr"
SpBtr=SpBtr[SpColumnNames]; SpBtr$Gtyp="Btr"
# All in one File:
   
df_list=list(crop=SpCrop,herb=SpHerb,shrub=SpShru,neddle=SpNtr,broad=SpBtr)
SpDF=do.call(rbind,df_list)
rownames(SpDF)=NULL
write.csv(SpDF,"SpeciationAll.csv")
rm(list=ls())
#===============================================================================
#
#Calculo el EF de cada "growth-type" (GT) agrupado por Ecotype

#Original Files
EF=read.csv("./orig/EFv210806.csv", sep=",");
SP=read.csv("./SpeciationAll.csv" , sep=",");

#Merge by VegID
M=merge(EF,SP,by = "VegID")
#Select only used columns
M[c("References","Comment","X","Family","GenusGroup","CommonName","Type","GrowthForm","VegID")]=NULL
#Multiply EF by Fraction and then Sum
efs=c(paste("EF",c(1:19),sep=""),paste("LDF",c(3:6),sep=""))
M[efs]=M[efs]*M$SpFrac

DF=aggregate(M[efs],by=list(gt=M$Gtyp,et=M$EcotypeID),FUN=sum)
DF=DF[order(DF$gt),]
DF=format(DF,digits=3,nsmall=3)
write.table(DF,"GtEFbyEcotype_new.csv",sep=",",dec=".",quote = F,row.names = F,col.names = F)
  
rm(list=ls())

File: src/db/calc_GtEFbyET_new.py
import pandas as pd
import numpy as np

# Set file paths
base_dir = "./EF_input_250205/"
ef_file = base_dir + "EFv250205.csv"

# Read original files
EF = pd.read_csv(ef_file)

SpCrop = pd.read_csv(base_dir + "SpeciationCrop241129.csv")
SpHerb = pd.read_csv(base_dir + "SpeciationHerb241129.csv")
SpShru = pd.read_csv(base_dir + "SpeciationShrub241129.csv")
SpTree = pd.read_csv(base_dir + "SpeciationTree241129.csv").iloc[:, :3]

# Add vegetation type labels
SpCrop["Gtyp"] = "Crop"
SpHerb["Gtyp"] = "Herb"
SpShru["Gtyp"] = "Shrub"
SpTree["Gtyp"] = "Tree"

SpColumnNames = ["EcotypeID", "VegID", "SpFrac", "Gtyp"]
SpCrop.columns = SpColumnNames
SpHerb.columns = SpColumnNames
SpShru.columns = SpColumnNames
SpTree.columns = SpColumnNames

## Merge GrowthForm from EF with Tree data
#df = pd.merge(EF[["VegID", "GrowthForm"]], SpTree, on="VegID", how="right")
#
## Process needleleaf trees (Ntr)
#SpNtr = df[df["GrowthForm"] == "Ntr"].copy()
#NtEtFracs = SpNtr.groupby("EcotypeID")["SpFrac"].sum().reset_index(name="total")
#for _, row in NtEtFracs.iterrows():
#    k = max(row["total"], 1e-23)
#    et = row["EcotypeID"]
#    SpNtr.loc[SpNtr["EcotypeID"] == et, "SpFrac"] /= k
#
## Process broadleaf trees (Btr)
#SpBtr = df[df["GrowthForm"] == "Btr"].copy()
#BtEtFracs = SpBtr.groupby("EcotypeID")["SpFrac"].sum().reset_index(name="total")
#for _, row in BtEtFracs.iterrows():
#    k = max(row["total"], 1e-23)
#    et = row["EcotypeID"]
#    SpBtr.loc[SpBtr["EcotypeID"] == et, "SpFrac"] /= k
#
## Prepare output for Ntr and Btr
#SpNtr = SpNtr[SpColumnNames]
#SpNtr["Gtyp"] = "Ntr"
#SpBtr = SpBtr[SpColumnNames]
#SpBtr["Gtyp"] = "Btr"

# Combine all speciation data
SpDF = pd.concat([SpCrop, SpHerb, SpShru, SpTree], ignore_index=True)
SpDF.to_csv("SpeciationAll.csv", index=False)

# ====================================
# Calculate emission factors (EF) grouped by growth type (Gtyp) and EcotypeID

# Reload input files
EF = pd.read_csv(ef_file)
SP = pd.read_csv("SpeciationAll.csv")

# Merge EF and SP
M = pd.merge(EF, SP, on="VegID")

# Drop unused columns
drop_cols = ["References", "Comment", "X", "Family", "GenusGroup", "CommonName", "Type", "GrowthForm"]
M = M.drop(columns=[col for col in drop_cols if col in M.columns])

# Define EF and LDF columns
efs = [f"EF{i}" for i in range(1, 20)] + [f"LDF{i}" for i in range(3, 7)]

# Multiply by speciation fraction
for col in efs:
    if col in M.columns:
        M[col] = M[col] * M["SpFrac"]

# Group and sum by Gtyp and EcotypeID
DF = M.groupby(["Gtyp", "EcotypeID"])[efs].sum().reset_index()
#DF = M.groupby(["EcotypeID"])[efs].sum().reset_index()
DF = DF.sort_values("Gtyp")
DF = DF.round(3)

# Write output file (no header or index)
DF.to_csv("GtEFbyEcotype_growthform.csv", sep=",", index=False, header=False, float_format="%.3f")

